blob/service.go

```diff
@@ -49,26 +50,131 @@ type Submitter interface {
 }
 
 type Service struct {
+	// ctx represents the Service's lifecycle context.
+	ctx    context.Context
+	cancel context.CancelFunc
 	// accessor dials the given celestia-core endpoint to submit blobs.
 	blobSubmitter Submitter
 	// shareGetter retrieves the EDS to fetch all shares from the requested header.
 	shareGetter share.Getter
 	// headerGetter fetches header by the provided height
 	headerGetter func(context.Context, uint64) (*header.ExtendedHeader, error)
+	// headerSub subscribes to new headers to supply to blob subscriptions.
+	headerSub func(ctx context.Context) (<-chan *header.ExtendedHeader, error)
+
+	// activeSubscriptions tracks the number of active subscriptions
+	activeSubscriptions sync.WaitGroup
+	stopped             bool
+	mu                  sync.Mutex
 }
 
 func NewService(
 	submitter Submitter,
 	getter share.Getter,
 	headerGetter func(context.Context, uint64) (*header.ExtendedHeader, error),
+	headerSub func(ctx context.Context) (<-chan *header.ExtendedHeader, error),
 ) *Service {
 	return &Service{
 		blobSubmitter: submitter,
 		shareGetter:   getter,
 		headerGetter:  headerGetter,
+		headerSub:     headerSub,
 	}
 }
 
+func (s *Service) Start(context.Context) error {
+	s.ctx, s.cancel = context.WithCancel(context.Background())
+	return nil
+}
+
+func (s *Service) Stop(context.Context) error {
+	s.mu.Lock()
+	if s.cancel != nil {
+		s.stopped = true
+		s.cancel()
+	}
+	s.mu.Unlock()
+
+	s.activeSubscriptions.Wait()
```

Comment:
What's the motivation behind this? A slightly retarded user may forget to close a subscription, which will block the node from stopping, and then he will start screaming about the node being wrong as it doesn't stop.

Reply:
Basically, why do we deprioritize the node lifecycle yet bring more complexity?

Reply:
Subscriptions will be terminated on canceling service context.

Reply:
So what's the point for waitGroup?

Reply:
Hmm, I see. I am not sure why we need that tbh,

Header comes every 12 seconds. The service context gets canceled on Stop instantaneously, stopping all the subscriptions that, in common case, sit still. Then they finish the waitgroup unblocking the Stop. How is that different from being disgraceful?

Reply:
The forcefull shutdown is 4 more lines of code, I can help to add it.

Reply:
We synced up on this and agreed to remove the graceful shutdown.